astropy/coordinates/transformations.py

```diff
@@ -746,8 +749,157 @@ def __call__(self, fromcoord, toframe):
+    def finite_difference_frameattr_name(self):
+        return self._finite_difference_frameattr_name
+    @finite_difference_frameattr_name.setter
+    def finite_difference_frameattr_name(self, value):
+        if value is None:
+            self._diff_attr_in_fromsys = self._diff_attr_in_tosys = False
+        else:
+            diff_attr_in_fromsys = value in self.fromsys.frame_attributes
+            diff_attr_in_tosys = value in self.tosys.frame_attributes
+            if diff_attr_in_fromsys or diff_attr_in_tosys:
+                self._diff_attr_in_fromsys = diff_attr_in_fromsys
+                self._diff_attr_in_tosys = diff_attr_in_tosys
+            else:
+                raise ValueError('Frame attribute name {} is not a frame '
+                                 'attribute of {} or {}'.format(value,
+                                                                self.fromsys,
+                                                                self.tosys))
+        self._finite_difference_frameattr_name = value
+
+    def __call__(self, fromcoord, toframe):
+        from .representation import (CartesianRepresentation,
+                                     CartesianDifferential)
+
+        supcall = self.func
+        if getattr(fromcoord.data, 'differentials', False):
+            # this is the finite difference case
+
+            if callable(self.finite_difference_dt):
+                dt = self.finite_difference_dt(fromcoord, toframe)
+            else:
+                dt = self.finite_difference_dt
+
+            fromdiffless = fromcoord.realize_frame(fromcoord.data.without_differentials())
+            reprwithoutdiff = supcall(fromdiffless, toframe)
+
+            # first we use the existing differential to compute an offset due to
+            # the already-existing velocity, but in the new frame
+            fromcoord_cart = fromcoord.cartesian
+            if self.symmetric_finite_difference:
+                halfdt = dt/2
+                fwdxyz = (fromcoord_cart.xyz +
+                          fromcoord_cart.differentials['s'].d_xyz*halfdt)
+                fwd = supcall(fromcoord.realize_frame(CartesianRepresentation(fwdxyz)), toframe)
+                backxyz = (fromcoord_cart.xyz -
+                           fromcoord_cart.differentials['s'].d_xyz*halfdt)
+                back = supcall(fromcoord.realize_frame(CartesianRepresentation(backxyz)), toframe)
+            else:
+                fwdxyz = (fromcoord_cart.xyz +
+                          fromcoord_cart.differentials['s'].d_xyz*dt)
+                fwd = supcall(fromcoord.realize_frame(CartesianRepresentation(fwdxyz)), toframe)
+                back = reprwithoutdiff
+            diffxyz = (fwd.cartesian - back.cartesian).xyz / dt
+
+            # now we compute the "induced" velocities due to any movement in the
+            # frame itself over time
+            attrname = self.finite_difference_frameattr_name
+            if attrname is not None:
```

Comment:
I feel like the logic here could be simplified, but nothing is coming immediately to mind and it shouldn't hold this up.